### PR TITLE
Public Suffix With SubDomain Handling

### PR DESCRIPTION
**Pull request by shabbir81**
New option added
Logic Updated for Public Suffix
Logic Added for Public Suffix with SubDomain

### CODE REVIEW

**Comment by limpkin** [comment]
@shabbir81 : warnings during compilation:
mooltifill\PublicSuffixManager.kt: (83, 33): 'toLowerCase(): String' is deprecated. Use lowercase() instead.
mooltifill\PublicSuffixManager.kt: (83, 72): 'toLowerCase(): String' is deprecated. Use lowercase() instead.
mooltifill\PublicSuffixManager.kt: (91, 20): Condition 'strippedTokens.last() != null' is always 'true'
mooltifill\PublicSuffixManager.kt: (156, 13): Condition 'result != null' is always 'true'

**Review by mathfactory — APPROVED** [review]

File: app/src/main/java/de/mathfactory/mooltifill/PublicSuffixManager.kt [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+                return tld
+            }
+            else -> { // Explicitly combining TLD+one part with next level subdomain
+                val strippedTokens = domain.substringBefore("."+tld).split(".")
+                val subdomain = strippedTokens.last() + "." + tld
+                return subdomain
+            }
+        }
+
+        return domain
```
Comment (mathfactory) [inline app/src/main/java/de/mathfactory/mooltifill/PublicSuffixManager.kt]: unreachable code

File: app/src/main/java/de/mathfactory/mooltifill/PublicSuffixManager.kt [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    fun skipPublicSuffix(domain: String): String
+    {
+        val isPublicSuffix = runBlocking { suffixList!!.isPublicSuffix(domain).await() }
+
+        if (isPublicSuffix)
+        {
+            return domain
+        }
+
+        return runBlocking { suffixList!!.stripPublicSuffix(domain).await() }!!
```
Comment (mathfactory) [inline app/src/main/java/de/mathfactory/mooltifill/PublicSuffixManager.kt]: Unnecessary non-null assertion (!!) on a non-null receiver of type String

File: app/src/main/java/de/mathfactory/mooltifill/PublicSuffixManager.kt [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+    var suffixList: PublicSuffixList? = null
+
+    /**
+     * Just initialize suffixList object with application provided context.
+     */
+
+    operator fun invoke(context: Context): PublicSuffixManager {
+        suffixList = PublicSuffixList(context)
+        suffixList!!.prefetch()
```
Comment (mathfactory) [inline app/src/main/java/de/mathfactory/mooltifill/PublicSuffixManager.kt]: 'Deferred' result is unused

**Review by mathfactory — CHANGES_REQUESTED** [review]
See comments

File: app/src/main/java/de/mathfactory/mooltifill/PublicSuffixManager.kt [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+ *
+ * Mooltifill is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Mooltifill.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
```
Comment (mathfactory) [inline app/src/main/java/de/mathfactory/mooltifill/PublicSuffixManager.kt]: Missing package de.mathfactory.mooltifill;

File: app/src/main/java/de/mathfactory/mooltifill/PublicSuffixManager.kt [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+import android.content.Context
+import kotlinx.coroutines.runBlocking
+import mozilla.components.lib.publicsuffixlist.PublicSuffixList
+
+object PublicSuffixManager {
+
+    var suffixList: PublicSuffixList? = null
```
Comment (mathfactory) [inline app/src/main/java/de/mathfactory/mooltifill/PublicSuffixManager.kt]: Property 'suffixList' could be private

**Comment by limpkin** [comment]
@shabbir81 could you update your PR so @mathfactory could review again?

**Comment by shabbir81** [comment]
Updated code and new PR opened

Commit Id 
9bcd8509800ad004c135b0a81f176a61b175da8a

File: app/src/main/java/de/mathfactory/mooltifill/PublicSuffixManager.kt [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+package de.mathfactory.mooltifill
+
+import android.content.Context
+import kotlinx.coroutines.runBlocking
+import mozilla.components.lib.publicsuffixlist.PublicSuffixList
+
+object PublicSuffixManager {
+
+    private lateinit var suffixList:PublicSuffixList
```
Comment (mathfactory) [inline app/src/main/java/de/mathfactory/mooltifill/PublicSuffixManager.kt]: suffixList is initialized in invoke(), which is called in SettingsActivity.onCreate(). It is not guaranteed, that this function is even called! (E.g. after reboot)

File: app/src/main/java/de/mathfactory/mooltifill/PublicSuffixManager.kt [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+    /**
+     * Just initialize suffixList object with application provided context.
+     */
+
+    operator fun invoke(context: Context): PublicSuffixManager {
+        suffixList = PublicSuffixList(context)
```
Comment (mathfactory) [inline app/src/main/java/de/mathfactory/mooltifill/PublicSuffixManager.kt]: the PublicSuffixManager object is a global singleton. Adding a reference to the context through PublicSuffixList(context) has issues: We have at least two different contexts in the Activity- and the Service-related part of the code. Also, storing a context over global lifetime of the application probably introduces a memory leak. Instead the context should better be passed to the get*() functions if possible (one way would be to add the context as a parameter to the SubstitutionPolicy.policies(query:String) function.

File: app/src/main/java/de/mathfactory/mooltifill/PublicSuffixManager.kt [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    /**
+     * Returns the public suffix and one more level; known as the registrable domain. Returns `null` if
+     * [domain] is a public suffix itself
+     * @param [domain] _must_ be a valid domain. [PublicSuffixList] performs no validation, and if any unexpected values
+     * are passed (e.g., a full URL, a domain with a trailing '/', etc) this may return an incorrect result.
+     */
+
+    fun getPublicSuffixPlusOne(domain: String): String
+    {
+        val isPublicSuffix = runBlocking { suffixList.isPublicSuffix(domain).await() }
```
Comment (mathfactory) [inline app/src/main/java/de/mathfactory/mooltifill/PublicSuffixManager.kt]: Please find a way to not use runBlocking { } as this negates the effect of async programming.

File: app/src/main/java/de/mathfactory/mooltifill/PublicSuffixManager.kt [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+        if (!isPublicSuffix)
+        {
+            val suffix = runBlocking { suffixList.getPublicSuffixPlusOne(domain).await() }
```
Comment (mathfactory) [inline app/src/main/java/de/mathfactory/mooltifill/PublicSuffixManager.kt]: Please find a way to not use runBlocking { } as this negates the effect of async programming.

File: app/src/main/java/de/mathfactory/mooltifill/PublicSuffixManager.kt [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    /**
+     * Returns the public suffix and two more level; Returns `null` if
+     * [domain] is a public suffix itself
+     * @param [domain] _must_ be a valid domain. [PublicSuffixList] performs no validation, and if any unexpected values
+     * are passed (e.g., a full URL, a domain with a trailing '/', etc) this may return an incorrect result.
+     */
+
+    fun getPublicSuffixWithSubdomain(domain: String): String
+    {
+        val isPublicSuffix = runBlocking { suffixList.isPublicSuffix(domain).await() }
```
Comment (mathfactory) [inline app/src/main/java/de/mathfactory/mooltifill/PublicSuffixManager.kt]: Please find a way to not use runBlocking { } as this negates the effect of async programming.

File: app/src/main/java/de/mathfactory/mooltifill/PublicSuffixManager.kt [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+        if (!isPublicSuffix)
+        {
+            val strippedDomain = domain.replace("www.","",true)
+            val tld = runBlocking { suffixList.getPublicSuffixPlusOne(domain).await() }
```
Comment (mathfactory) [inline app/src/main/java/de/mathfactory/mooltifill/PublicSuffixManager.kt]: Please find a way to not use runBlocking { } as this negates the effect of async programming.

File: app/src/main/java/de/mathfactory/mooltifill/PublicSuffixManager.kt (hunk 6 above)
Comment (mathfactory) [reply app/src/main/java/de/mathfactory/mooltifill/PublicSuffixManager.kt]: (I expect an application crash after reboot, when accessing credentials without going to the settings first)

File: app/src/main/res/xml/root_preferences.xml [hunk 12]
```diff
@@ -35,7 +35,7 @@
             app:title="Subdomain Handling Behaviour"
             app:entries="@array/www_substitution_entries"
             app:entryValues="@array/www_substitution_values"
-            app:defaultValue="ApplyPublicSuffixList"
+            app:defaultValue="PublicSuffix"
```
Comment (mathfactory) [inline app/src/main/res/xml/root_preferences.xml:38]: If you change the default value here, please also change it in SettingsActivity:100, which is currently "UrlSubstitutionPolicies.RemoveWww". There are two cases which lead to using a default and they are handled differently, sorry for this ;)

**Review by mathfactory — CHANGES_REQUESTED** [review]

**Comment by mathfactory** [comment]
I hope my comments do not sound too "harsh", thank you very much for the constribution! ;) I think best way would be to adapt the SubstitutionPolicy classes to be async/suspend and to pass a context in fun policies(). I can also offer to have a look into this, just give me a hint.

File: app/src/main/res/xml/root_preferences.xml (hunk 12 above)
Comment (shabbir81) [reply app/src/main/res/xml/root_preferences.xml:38]: sure

Comment (shabbir81) [reply app/src/main/res/xml/root_preferences.xml:38]: Updated

**Pull request merged**